src/rgw/rgw_lc.cc

```diff
@@ -1390,3 +1397,132 @@ int RGWLC::remove_bucket_config(RGWBucketInfo& bucket_info,
+
+  const auto& aiter = bucket_attrs.find(RGW_ATTR_LC);
+  if (aiter == bucket_attrs.end())
+    return hdr;
+
+  bufferlist::const_iterator iter{&aiter->second};
+  try {
+      config.decode(iter);
+  } catch (const buffer::error& e) {
+      ldpp_dout(dpp, 0) << __func__
+			<<  "() decode life cycle config failed"
+			<< dendl;
+      return hdr;
+  } /* catch */
+
+  boost::optional<ceph::real_time> expiration_date;
+  boost::optional<std::string> rule_id;
+
+  const auto& rule_map = config.get_rule_map();
+  for (const auto& ri : rule_map) {
+    const auto& rule = ri.second;
+    auto& id = rule.get_id();
+    auto& prefix = rule.get_prefix();
+    auto& filter = rule.get_filter();
+    auto& expiration = rule.get_expiration();
+    auto& noncur_expiration = rule.get_noncur_expiration();
+
+    ldpp_dout(dpp, 10) << "rule: " << ri.first
+		       << " prefix: " << prefix
+		       << " expiration: "
+		       << " date: " << expiration.get_date()
+		       << " days: " << expiration.get_days()
+		       << " noncur_expiration: "
+		       << " date: " << noncur_expiration.get_date()
+		       << " days: " << noncur_expiration.get_days()
+		       << dendl;
+
+    /* skip if rule !enabled
+     * if rule has prefix, skip iff object !match prefix
+     * if rule has tags, skip iff object !match tags
+     * note if object is current or non-current, compare accordingly
+     * if rule has days, construct date expression and save iff older
+     * than last saved
+     * if rule has date, convert date expression and save iff older
+     * than last saved
+     * if the date accum has a value, format it into hdr
+     */
+
+    if (! rule.is_enabled())
+      continue;
+
+    if(! prefix.empty()) {
+      if (! boost::starts_with(obj_key.name, prefix))
+	continue;
+    }
+
+    if (filter.has_tags()) {
+      bool tag_match = true;
+      const RGWObjTags& rule_tagset = filter.get_tags();
+      RGWObjTags::tag_map_t obj_tag_map = obj_tagset.get_tags();
+      for (auto& tag : rule_tagset.get_tags()) {
+	if (obj_tag_map.find(tag.first) == obj_tag_map.end()) {
+	  tag_match = false;
+	  break;
+	}
+      }
+      if (! tag_match)
+	continue;
+    }
+
+    // compute a uniform expiration date
+    boost::optional<ceph::real_time> rule_expiration_date;
+    const LCExpiration& rule_expiration =
+      (obj_key.instance.empty()) ? expiration : noncur_expiration;
+
+    if (rule_expiration.has_date()) {
+      rule_expiration_date =
+	boost::optional<ceph::real_time>(
+	  ceph::from_iso_8601(rule.get_expiration().get_date()));
+      rule_id = boost::optional<std::string>(id);
+    } else {
+      if (rule_expiration.has_days()) {
+	rule_expiration_date =
+	  boost::optional<ceph::real_time>(
+	    mtime + make_timespan(rule_expiration.get_days()*24*60*60));
+	rule_id = boost::optional<std::string>(id);
+      }
+    }
+
+    // update earliest expiration
+    if (rule_expiration_date) {
+      if ((! expiration_date) ||
+	  ((expiration_date &&
```

Comment:
@mattbenjamin no need to check here for `expiration_data` not empty, already checked it in previous line

Reply:
unless you wanted to check that it's not zero (in which case need `*expiration_date`)

Reply:
no, this was not a check for 0

Reply:
ok, rule_expiration_date may not have been assigned in either branch of the prior has_date()/has_days() conditional.  meanwhile, expiration_date is a default-initialized boost::optional declared at l. 1435 (my updated source, and may not have been updated at l. 1509 (my updated source).  So we must check whether it was initialized in the preceding logic.   Checking the boost::optional seems not-optional therefore...

Reply:
ok, never mind, I kept eye skipping it